### PR TITLE
[storage] Add some tests for storage primitives

### LINKED ISSUE #403: Add more unit tests for storage2 primitive implementations
One of the few unit tests that have been left out of the initial implementation of the `storage2` module have been unit tests that check if the provided implementation of all Rust primitives, arrays, tuples and prelude types is actually correct.

These implementations are critical enough that they should not be taken for granted.
This issue is about writing unit tests for their `SpreadLayout` and `PackedLayout` implementations.

A show case for unit tests of all Rust primitives such as `i32` and also ink! primitives such as `Key` does already exist and can be taken as inspiration.

## ToDo List

- [ ] Get familiar with writing unit tests using the already existing unit tests for ink! and Rust primitives for `SpreadLayout` and `PackedLayout`.
    - Find them here: https://github.com/paritytech/ink/blob/master/core/src/storage2/traits/impls/prims.rs#L236
- [ ] Write unit tests for Rust arrays.
- [ ] Write unit tests for Rust tuples.
- [ ] Write unit tests for Rust's `Option` and `Result` types.
- [ ] Write unit tests for the rest prelude types, e.g. `Box`, `BTreeMap`, etc.

### PR DESCRIPTION
**Pull request by cmichi**
Closes #403.

### CODE REVIEW

File: crates/storage/src/traits/impls/fuzz_tests.rs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+///  Receives a slice, returns an array.
+fn clone_into_array<A, T>(slice: &[T]) -> A
+where
+    A: Default + AsMut<[T]>,
+    T: Clone,
+{
+    let mut a = A::default();
+    <A as AsMut<[T]>>::as_mut(&mut a).clone_from_slice(slice);
+    a
+}
```
Comment (Robbepop) [inline crates/storage/src/traits/impls/fuzz_tests.rs:32]: Pretty cool!

File: crates/storage/src/traits/impls/fuzz_tests.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+/// Tests if a fuzzed `[i32; 32]` array results in the same object when
+/// pushed/pulled from storage (for `spread` and `packed`).
+#[quickcheck]
+fn fuzz_pull_push_pull_array(x: Vec<i32>) -> TestResult {
+    if x.len() < 32 {
+        return TestResult::discard()
+    }
```
Comment (Robbepop) [inline crates/storage/src/traits/impls/fuzz_tests.rs:40]: Why discard in this case?

File: crates/storage/src/traits/impls/fuzz_tests.rs (hunk 1 above)
Comment (Robbepop) [reply crates/storage/src/traits/impls/fuzz_tests.rs:32]: However, why not use `Vec<u8>`? ^^

File: crates/storage/src/traits/impls/fuzz_tests.rs (hunk 2 above)
Comment (Robbepop) [reply crates/storage/src/traits/impls/fuzz_tests.rs:40]: If we used `Vec<_>` we were fine.

**Review by Robbepop — COMMENTED** [review]
All in all looks good to me. Some questions though.

File: crates/storage/src/traits/impls/fuzz_tests.rs (hunk 2 above)
Comment (cmichi) [reply crates/storage/src/traits/impls/fuzz_tests.rs]: > Why discard in this case?

I added a clarifying comment.

**Review by Robbepop — APPROVED** [review]
LGTM

**Pull request merged**